src/main/java/wooteco/subway/domain/Sections.java

```diff
@@ -0,0 +1,215 @@
+            values.add(0, newSection);
+            return newSection;
+        }
+
+        if (getDownDestination().equals(newSection.getUpStation())) {
+            values.add(newSection);
+            return newSection;
+        }
+
+        for (Section section : values) {
+            if (section.hasSameUpStation(newSection)) {
+                return insertAfterUpStation(newSection, section);
+            }
+
+            if (section.hasSameDownStation(newSection)) {
+                return insertBeforeDownStation(newSection, section);
+            }
+        }
+        throw new IllegalStateException("구간을 추가할 수 있는 상태가 아닙니다.");
+    }
+
+    private Section insertAfterUpStation(Section newSection, Section section) {
+        int index = values.indexOf(section);
+        Section sectionToModify = new Section(section.getId(), newSection.getDownStation(), section.getDownStation(),
+            getNewDistance(section, newSection));
+        values.set(index, newSection);
+        values.add(index + 1, sectionToModify);
+        return sectionToModify;
+    }
+
+    private int getNewDistance(Section section, Section newSection) {
+        int newDistance = section.getDistance() - newSection.getDistance();
+        if (newDistance <= 0) {
+            throw new IllegalArgumentException("추가하려는 구간의 거리가 기존 구간의 거리보다 작아야합니다.");
+        }
+        return newDistance;
+    }
+
+    private Section insertBeforeDownStation(Section newSection, Section section) {
+        int index = values.indexOf(section);
+        Section sectionToModify = new Section(section.getId(), section.getUpStation(), newSection.getUpStation(),
+            getNewDistance(section, newSection));
+        values.set(index, sectionToModify);
+        values.add(index + 1, newSection);
+        return sectionToModify;
+    }
+
+    private void validateAddable(Section section) {
+        List<Section> foundSections = findSectionsOverlapped(section);
+
+        if (isInvalid(section, foundSections)) {
+            throw new IllegalArgumentException("이미 포함된 두 역을 가진 Section 을 추가할 수 없습니다.");
+        }
+    }
+
+    private List<Section> findSectionsOverlapped(Section newSection) {
+        return values.stream()
+            .filter(section ->
+                section.contains(newSection.getDownStation()) || section.contains(newSection.getUpStation()))
+            .collect(Collectors.toList());
+    }
+
+    private boolean isInvalid(Section section, List<Section> foundSections) {
+        return hasSameStations(section) ||
+            isInvalidCountInMiddle(foundSections) || isSameWithDestinations(foundSections);
+    }
+
+    private boolean hasSameStations(Section newSection) {
+        return values.stream().anyMatch(section -> section.hasSameStations(newSection));
+    }
+
+    private boolean isInvalidCountInMiddle(List<Section> foundSections) {
+        int count = foundSections.size();
+        return count < 1 || count > 2;
```

Comment:
이거 결국엔 1 또는 2여야 한다는 거죠? 뭔가 헷갈리네요 ㅎㅎ

Reply:
남겨주신 피드백을 보고 부등호보다 일치 여부로 비교하는 게 더 간단하겠다고 판단하여 수정하였으며, 의미 있는 상수로 추출하여보았습니다👍 감사합니다.
```suggestion
        private boolean isInvalidCountInMiddle(List<Section> foundSections) {
        int count = foundSections.size();
        return count != OVERLAP_COUNT_ADD_DESTINATION && count != OVERLAP_COUNT_ADD_MIDDLE;
    }
```